### PR TITLE
LinearLayout conversion interface

### LINKED ISSUE #1874: Generic LinearLayout conversion interface.
Currently, our DPAS->LL conversion code is outside upstream's `toLinearLayout` function, so that we have to manually detect and call our `DPAStoLinearLayout` whenever DPAS->LL conversion is needed.
We need to create a generic interface that allows implementing LinearLayout conversion outside the upstream code.

### PR DESCRIPTION
**Pull request by hwnam831**
This PR closes #1874 
Instead of detecting and calling `DPAStoLinearLayout` whenever it is needed, now we can use generic `toLinearLayout` for DPASEncodingAttr layouts.
Also fixed the DPAS->LL logic bug and updated the LIT/unit tests accordingly.

The interface change will be upstreamed after merging(#1785)

### CODE REVIEW

**Review by chengjunlu — DISMISSED** [review]

File: include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td [hunk 1]
```diff
@@ -799,6 +799,21 @@ def MmaEncodingTrait : AttrInterface<"MmaEncodingTrait"> {
   ];
 }
 
+//===----------------------------------------------------------------------===//
```
Comment (etiotto) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td:802]: Adding this change downstream will cause us to deviate from upstream code more. Please post a PR upstream to introduce this change.

**Review by whitneywhtsang — COMMENTED** [review]
LGTM, let's attempt to upstream this change, thanks!

File: include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td [hunk 2]
```diff
@@ -799,6 +799,21 @@ def MmaEncodingTrait : AttrInterface<"MmaEncodingTrait"> {
...
+def LinearLayoutTrait : AttrInterface<"LinearLayoutTrait"> {
+  let cppNamespace = "::mlir::triton::gpu";
+  let methods = [
+    InterfaceMethod<"Convert to LinearLayout.",
+                    "LinearLayout",
+                    "toLinearLayout",
+                    (ins
+                    "ArrayRef<int64_t>":$shape,
+                    "std::optional<int32_t>":$elemBitWidth)>
```
Comment (victor-eds) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td:813]:
```suggestion
                    (ins "ArrayRef<int64_t>":$shape,
                         "std::optional<int32_t>":$elemBitWidth)>
```
NIT

File: test/TritonIntelGPU/tritonintlgpu-nested-layout.mlir [hunk 3]
```diff
@@ -11,46 +11,32 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 4 :
...
+    // CHECK:           %[[OFFSET_X_0:.*]] = llvm.xor %[[VAL_44]], %[[CST_0]] : i32
+    // CHECK-DAG:           %[[OFFSET_Y_0:.*]] = llvm.xor %[[VAL_40]], %[[CST_0]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_1:.*]] = llvm.xor %[[VAL_44]], %[[CST_2]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_2:.*]] = llvm.xor %[[VAL_44]], %[[CST_4]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_3:.*]] = llvm.xor %[[VAL_44]], %[[CST_6]] : i32
+    // CHECK-DAG:           %[[OFFSET_Y_1:.*]] = llvm.xor %[[VAL_40]], %[[CST_16]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_4:.*]] = llvm.xor %[[VAL_44]], %[[CST_16]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_5:.*]] = llvm.xor %[[VAL_44]], %[[CST_18]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_6:.*]] = llvm.xor %[[VAL_44]], %[[CST_20]] : i32
+    // CHECK-DAG:           %[[OFFSET_X_7:.*]] = llvm.xor %[[VAL_44]], %[[CST_22]] : i32
```
Comment (victor-eds) [inline test/TritonIntelGPU/tritonintlgpu-nested-layout.mlir:39]: NIT: I think we can remove `-DAG` to simplify the tests as we know the operation generation order here.

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/Attributes.h [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #ifndef TRITON_DIALECT_TRITON_INTEL_GPU_IR_ATTRIBUTES_H
 #define TRITON_DIALECT_TRITON_INTEL_GPU_IR_ATTRIBUTES_H
 
+#include "triton/Dialect/TritonGPU/IR/Attributes.h"
 #define GET_ATTRDEF_CLASSES
 #include "intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUAttrDefs.h.inc"
```
Comment (victor-eds) [inline third_party/intel/include/Dialect/TritonIntelGPU/IR/Attributes.h:6]:
```suggestion
#include "triton/Dialect/TritonGPU/IR/Attributes.h"

#define GET_ATTRDEF_CLASSES
#include "intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUAttrDefs.h.inc"
```
NIT

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 5]
```diff
@@ -415,6 +416,12 @@ void DpasEncodingAttr::print(AsmPrinter &printer) const {
           << "}>";
 }
 
+LinearLayout
+DpasEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape,
+                                 std::optional<int32_t> elemBitWidth) const {
+  return DPAStoLinearLayout(shape, *this, 2 /*opIdx*/);
+}
```
Comment (victor-eds) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:423]:
```suggestion
LinearLayout
DpasEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape,
                                 std::optional<int32_t> elemBitWidth) const {
  return DPAStoLinearLayout(shape, *this);
}
```

NIT: Can we exploit default value? If not, I think `/*opIdx=*/2` is preferred (at least clang-format likes it better)

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp [hunk 6]
```diff
@@ -547,36 +558,26 @@ DPAStoLinearLayout(ArrayRef<int64_t> shape, Attribute layout, unsigned opIdx) {
...
+        LinearLayout::identity1D(warpsPerCTA[0], kWarp, outDimNames[0]);
     nonKDim = 0;
     KDim = 1;
   }
 
+  // Lastly, the layout repeats to match the shape.
   // Operand A/B repeats through the K-dimension first then repeats
-  // through non-K dimension.
+  // through the non-K dimension.
+  SmallVector<int64_t> numReps = dpas.getDPASRepetitions(shape, opIdx);
```
Comment (victor-eds) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp:574]:
```suggestion
  ArrayRef<int64_t> numReps = dpas.getDPASRepetitions(shape, opIdx);
```

NIT

Comment (hwnam831) [reply third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp:574]: The return type of `getDPASRepetitions` is `SmallVector<int64_t>`. Is there a reason why you prefer `ArrayRef`?

File: include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td (hunk 1 above)
Comment (hwnam831) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td:802]: I will try to upstream this part (#1785).

File: test/TritonIntelGPU/tritonintlgpu-nested-layout.mlir (hunk 3 above)
Comment (hwnam831) [reply test/TritonIntelGPU/tritonintlgpu-nested-layout.mlir]: Modified.

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/Attributes.h (hunk 4 above)
Comment (hwnam831) [reply third_party/intel/include/Dialect/TritonIntelGPU/IR/Attributes.h]: Done

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp (hunk 5 above)
Comment (hwnam831) [reply third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp]: Done

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp (hunk 6 above)
Comment (victor-eds) [reply third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp:574]: Oh, I thought it was a reference. LGTM.

**Review by victor-eds — APPROVED** [review]

**Review by chengjunlu — APPROVED** [review]

**Comment by etiotto** [comment]
Because this change increases our differences from upstream code we can change it to a draft and resume the work once the code changes in common files have been merged upstream.

**Comment by whitneywhtsang** [comment]
Merged in https://github.com/intel/intel-xpu-backend-for-triton/pull/1942.

**Pull request closed without merging**